### PR TITLE
Add new sample for graphics pipeline libraries

### LINKED ISSUE #441: Add sample for VK_EXT_graphics_pipeline_library
This issue will be used to keep track for adding a new sample for the `VK_EXT_graphics_pipeline_library` extensions. See https://github.com/KhronosGroup/Vulkan-Docs/issues/1808

This will be a basic api sample that shows how to use the new extension.

### PR DESCRIPTION
**Pull request by SaschaWillems**
## Description

This PR adds a sample for the new `VK_EXT_graphics_pipeline_library` extension. The sample creates a pipeline library for shared pipeline state parts and reuses this to generate randomized fragment shaders in a background thread. The sample also includes a small tutorial.

I have tested the sample on Windows 10 with an NVIDIA RTX 2060 and recent Vulkan developer driver.

Fixes #441

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

### CODE REVIEW

**Comment by marty-johnson59** [comment]
Added reviewer suggestions.  Would be great if we could get the DT folks (@aurolou. @mklefrancois) to review as I believe that's the target for this EXT.  Thanks!

**Review by gary-sweet — COMMENTED** [review]
Builds ok for me, and correctly reports that the required extensions are not supported, but that's all I can test I'm afraid.

**Review by gpx1000 — DISMISSED** [review]
works well here Linux

File: samples/README.md [hunk 1]
```diff
@@ -227,6 +227,10 @@ sample rate map as a base for the next frame.
 Render a simple scene showing the basics of ray tracing, including reflection and shadow rays. The sample creates some geometries and create a bottom acceleration structure for each, then make instances of those, using different materials and placing them at different locations.
 <br/>
 
+### [Graphics pipeline library](./extensions/graphics_pipeline_library)<br/>
+**Extension**: [```VK_EXT_graphics_pipeline_library```](https://www.khronos.org/registry/vulkan/specs/1.3-extensions/man/html/VK_EXT_graphics_pipeline_library.html) <br/>
+Uses the graphics pipeline library extensions to improve run-time pipeline creation. Instead of creating the whole pipeline at once, this sample makes use of that extension to pre-built shared pipeline parts like like vertex input state and fragment output state. These building blocks are then used to create pipelines at runtime, improving build times compared to traditional pipeline creation.
```
Comment (TomAtkinsonArm) [inline samples/README.md]: `pre-built shared pipeline parts like like vertex input state`

to

`pre-build shared pipeline parts like vertex input state`

**Review by TomAtkinsonArm — COMMENTED** [review]
Small typo. Extension not supported, I will update graphics drivers and try again

File: samples/README.md (hunk 1 above)
Comment (SaschaWillems) [reply samples/README.md]: Thanks for spotting this. I have pushed a change to fix this.

**Review by gpx1000 — DISMISSED** [review]

File: samples/extensions/graphics_pipeline_library/README.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+executable_pipeline_create_info.pNext = &linking_info;
+executable_pipeline_create_info.flags = VK_PIPELINE_CREATE_LINK_TIME_OPTIMIZATION_BIT_EXT;
+
+VkPipeline executable = VK_NULL_HANDLE;
+vkCreateGraphicsPipelines(get_device().get_handle(), thread_pipeline_cache, 1, &executable_pipeline_create_info, nullptr, &executable);
+```
+
+This will result in the pipeline state object to be used at draw time. 
+
+A note on `VK_PIPELINE_CREATE_LINK_TIME_OPTIMIZATION_BIT_EXT`: This is an optimization flag. If specified, implementations are allowed to do additional optimization passes. This may increase build times but can in turn result in lower runtime costs.
```
Comment (Tobski) [inline samples/extensions/graphics_pipeline_library/README.md:125]: I'd suggest it's worth mentioning here that we suggest apps build first without this flag, and then build again with this flag enabled on the slow path, and switch it out for the fast-linked version when available.

File: samples/extensions/graphics_pipeline_library/graphics_pipeline_library.cpp [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+	VK_CHECK(vkAllocateDescriptorSets(get_device().get_handle(), &alloc_info, &descriptor_set));
+
+	VkDescriptorBufferInfo            uniform_buffer_descriptor = create_descriptor(*uniform_buffer);
+	std::vector<VkWriteDescriptorSet> write_descriptor_sets     = {
+        vkb::initializers::write_descriptor_set(descriptor_set, VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER, 0, &uniform_buffer_descriptor),
+    };
+	vkUpdateDescriptorSets(get_device().get_handle(), static_cast<uint32_t>(write_descriptor_sets.size()), write_descriptor_sets.data(), 0, nullptr);
+}
+
+// Compiling shaders can be simplified with the new extension, so we only require code to generate the SPIR-V in thhis sample
```
Comment (Tobski) [inline samples/extensions/graphics_pipeline_library/graphics_pipeline_library.cpp]:
```suggestion
// Compiling shaders can be simplified with the new extension, so we only require code to generate the SPIR-V in this sample
```

File: samples/extensions/graphics_pipeline_library/README.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+This sample demonstrates that functionality by creating the shared vertex input interface, pre-rasterization shader state and fragment output interface parts only once up-front, and then re-uses them to create pipelines with customized fragment shaders using random lighting models at runtime. Pipelines are created in a background thread and once they're created, command buffers are updated to display a mesh using the new pipeline.
+
+## Additional resources
+
+* [Reducing Draw Time Hitching with VK_EXT_graphics_pipeline_library](https://www.khronos.org/blog/reducing-draw-time-hitching-with-vk-ext-graphics-pipeline-library)
+* [Extension proposal](https://github.com/KhronosGroup/Vulkan-Docs/blob/main/proposals/VK_EXT_graphics_pipeline_library.asciidoc)
+
+## Conclusion
+
+With the new extension it's now possible to separate the monolithic pipeline state into multiple parts that can be r-used and built independently. This opens up new possibilities for optimizing pipeline creation and reducing hick-ups at runtime.
```
Comment (Tobski) [inline samples/extensions/graphics_pipeline_library/README.md]:
```suggestion
With the new extension it's now possible to separate the monolithic pipeline state into multiple parts that can be reused and built independently. This opens up new possibilities for optimizing pipeline creation and reducing hitches at runtime.
```

Hitches is usually the term used by most ISVs I've heard.

File: samples/extensions/graphics_pipeline_library/README.md (hunk 2 above)
Comment (Tobski) [reply samples/extensions/graphics_pipeline_library/README.md:125]: I also wonder if it's worth making the sample behave like this? The pieces to do it are all there, so I don't think it'd be a big change.

**Review by Tobski — CHANGES_REQUESTED** [review]
Mostly LGTM, some minor comments and one suggested modification.

File: samples/extensions/graphics_pipeline_library/README.md (hunk 2 above)
Comment (SaschaWillems) [reply samples/extensions/graphics_pipeline_library/README.md:125]: That's indeed a good point. I don't want the sample to become too complicated, so would it be okay if there was a checkbox in the UI where the optimization flag can be toggled? That would let people actually see the difference in build time very easily.

Comment (Tobski) [reply samples/extensions/graphics_pipeline_library/README.md:125]: That's probably a reasonable compromise.

**Comment by SaschaWillems** [comment]
@Tobski: Thank you very much for your valuable feedback. I have updated the sample and added an UI toggle for the link time optimization flag and also fixed the readme.

File: samples/extensions/graphics_pipeline_library/graphics_pipeline_library.cpp [hunk 5]
```diff
@@ -0,0 +1,539 @@
...
+
+	// Link the library parts into a graphics pipeline
+	VkPipelineLibraryCreateInfoKHR linking_info{};
+	linking_info.sType        = VK_STRUCTURE_TYPE_PIPELINE_LIBRARY_CREATE_INFO_KHR;
+	linking_info.libraryCount = static_cast<uint32_t>(libraries.size());
+	linking_info.pLibraries   = libraries.data();
+
+	VkGraphicsPipelineCreateInfo executable_pipeline_create_info{};
+	executable_pipeline_create_info.sType = VK_STRUCTURE_TYPE_GRAPHICS_PIPELINE_CREATE_INFO;
+	executable_pipeline_create_info.pNext = &linking_info;
```
Comment (pdaniell-nv) [inline samples/extensions/graphics_pipeline_library/graphics_pipeline_library.cpp]: executable_pipeline_create_info.layout must not be left NULL see https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-VkGraphicsPipelineCreateInfo-flags-06729 and https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-VkGraphicsPipelineCreateInfo-flags-06730. This came about due a recent spec change: https://gitlab.khronos.org/vulkan/vulkan/-/merge_requests/5164

**Review by pdaniell-nv — CHANGES_REQUESTED** [review]

File: samples/README.md [hunk 6]
```diff
@@ -235,6 +235,10 @@ MoltenVk: [MoltenVk](https://github.com/KhronosGroup/MoltenVK).  Also demonstrat
 for querying which features of the full Vulkan spec are not currently supported by the non-conformant Vulkan 
 implementation.
 
+### [Graphics pipeline library](./extensions/graphics_pipeline_library)<br/>
+**Extension**: [```VK_EXT_graphics_pipeline_library```](https://www.khronos.org/registry/vulkan/specs/1.3-extensions/man/html/VK_EXT_graphics_pipeline_library.html) <br/>
+Uses the graphics pipeline library extensions to improve run-time pipeline creation. Instead of creating the whole pipeline at once, this sample makes use of that extension to pre-build shared pipeline parts like like vertex input state and fragment output state. These building blocks are then used to create pipelines at runtime, improving build times compared to traditional pipeline creation.
```
Comment (TomAtkinsonArm) [inline samples/README.md]: `like like` -> `such as`

**Review by TomAtkinsonArm — CHANGES_REQUESTED** [review]
LGTM - A tiny typo I missed earlier. Cant run as my devices do not support the extension

File: samples/extensions/graphics_pipeline_library/graphics_pipeline_library.cpp (hunk 5 above)
Comment (SaschaWillems) [reply samples/extensions/graphics_pipeline_library/graphics_pipeline_library.cpp]: Thanks for pointing that out. I have updated the sample accordingly.

File: samples/README.md (hunk 6 above)
Comment (SaschaWillems) [reply samples/README.md]: Thanks for pointing this out. Fixed it.

**Review by pdaniell-nv — APPROVED** [review]

**Comment by marty-johnson59** [comment]
@SaschaWillems, this has now  had approvals from Tobias, Piers and Steve (tho some are stale).  Looks like all threads are resolved now?  Are we ready to merge?

**Comment by SaschaWillems** [comment]
Yes, feel free to merge :)

**Pull request merged**